src/GafferScene/SetQuery.cpp

```diff
@@ -0,0 +1,439 @@
+		matchesInternalPlug()->hash( h );
+	}
+	else
+	{
+		ComputeNode::hash( output, context, h );
+	}
+}
+
+void SetQuery::compute( Gaffer::ValuePlug *output, const Gaffer::Context *context ) const
+{
+	if( output == matchesInternalPlug() )
+	{
+		static_cast<ObjectPlug *>( output )->setValue( computeMatchesInternal( context ) );
+	}
+	else if( output == matchesPlug() )
+	{
+		const ScenePlug::ScenePath location = ScenePlug::stringToPath( locationPlug()->getValue() );
+		ScenePlug::PathScope pathScope( context, &location );
+		auto matchesData = boost::static_pointer_cast<const MatchesData>( matchesInternalPlug()->getValue() );
+		static_cast<StringVectorDataPlug *>( output )->setValue( matchesData->matches );
+	}
+	else if( output == firstMatchPlug() )
+	{
+		const ScenePlug::ScenePath location = ScenePlug::stringToPath( locationPlug()->getValue() );
+		ScenePlug::PathScope pathScope( context, &location );
+		auto matchesData = boost::static_pointer_cast<const MatchesData>( matchesInternalPlug()->getValue() );
+		static_cast<StringPlug *>( output )->setValue( matchesData->matches->readable().size() ? matchesData->matches->readable().front() : "" );
+	}
+	else
+	{
+		ComputeNode::compute( output, context );
+	}
+}
+
+Gaffer::ValuePlug::CachePolicy SetQuery::computeCachePolicy( const ValuePlug *output ) const
+{
+	if( output == firstMatchPlug() || output == matchesPlug() )
+	{
+		// Since these computes just delegate directly to `__matchesInternal` there's
+		// no benefit in caching the results separately.
+		return ValuePlug::CachePolicy::Uncached;
+	}
+	else
+	{
+		return ComputeNode::computeCachePolicy( output );
+	}
+}
+
+bool SetQuery::affectsMatchesInternal( const Gaffer::Plug *input ) const
+{
+	return
+		input == scenePlug()->setNamesPlug() ||
+		input == setsPlug() ||
+		input == scenePlug()->setPlug() ||
+		input == inheritPlug()
+	;
+}
+
+void SetQuery::hashMatchesInternal( const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	ComputeNode::hash( matchesInternalPlug(), context, h );
+
+	auto path = context->getIfExists<ScenePlug::ScenePath>( ScenePlug::scenePathContextName );
+	if( !path )
+	{
+		for( const auto &setName : matchingSetNames( scenePlug(), setsPlug()->getValue() ) )
+		{
+			h.append( setName );
+			// Note : We don't actually compute the set in the corresponding
+			// part of `computeMatchesInternal()` so don't _need_ to hash it
+			// here. But doing it here means we only pay for it once, and get
+			// to inherit it down to all descendants far more cheaply than we
+			// would by repeating it for every location.
+			h.append( scenePlug()->setHash( setName ) );
+		}
+		inheritPlug()->hash( h );
+	}
+	else
+	{
+		ParentScope scope( context, *path );
+		matchesInternalPlug()->hash( h );
+		h.append( *path );
+		// Note : Here we _should_ be hashing the relevant sets, but instead we
+		// rely on inheriting the hash via the recursive inclusion of
+		// `matchesInternalPlug()->hash( h )`.
```

Comment:
I think you're right, but I'd like to keep things as-is for two reasons :

1. It's hard enough to keep hash/compute in-sync as it is, so any deviations from a one-to-one match need strong justification.
2. I tried it just now, and didn't get any performance benefit in `testScaling()`. If anything (I'm not sure how much noise there is in the measurement), I got a 5% slowdown (or 20% if I take one extra-performant outlier from the baseline case). My hypothesis is that this is because we're going to need the hash for the parent location (rather than the root) anyway, to allow us to retrieve its MatchesData value in `compute()`. So it's better to access that twice than it is to yoyo between the parent hash and the root hash.

Reply:
That explanation kinda makes sense ... no reason to change if this is faster.